### ISSUE TITLE
Display XP-to-advance

### ISSUE THREAD

**Issue by jostephd** [wesnoth/wesnoth#3974]
After posting https://github.com/wesnoth/wesnoth/issues/1294#issuecomment-469665282 I came up with the idea of changing the XP display, like this:

![screenshot_2019-03-05_19-20-17](https://user-images.githubusercontent.com/24784687/54036621-fa8ce580-41b3-11e9-8be7-6c49dac0287a.png)

The screenshot shows a fresh recruit. Only blue part is changed, in master it reads "0/43XP". A unit with 42/43XP would show as "1↓" with this change. In general, the blue number is how many XP are required to advance, for example, "10" for a Red Mage with 90/100XP.

Received mixed feedback on discord, in particular, @ProditorMagnus's XP mod uses a unit's accumulated experience as spendable resources so kinda relies on having a unit's accumulated XP shown in the UI somewhere. However, for SP/MP games that don't use such mods, the accumulated XP is irrelevant, and showing it in the UI encourages players to make the sunk cost fallacy.

I have implemented this for my own use, see spoiler. It is purely a presentation layer change, there's no change to `experience` and `max_experience` properties of units and unit types. This issue is about whether to add this to mainline with a way to enable/disable it (a preference, or a Lua API to the effect of "I use accumulated XP as a resource", or something).

<details> <summary>The patch I use</summary>

I wrote this patch for my own use so it uses a hidden preference. That part will have to change before inclusion in mainline.

```
commit 1ee433eaaa8715782db288a1c1333908342ed4f1
Author: josteph <josteph@fastmail.com>
Date:   Tue Mar 5 19:43:35 2019 +0000

    Change experience display in UI from "0/42XP" to 42↓, with an overriding hidden preference for Ravana's XP mod and such.

diff --git a/src/gui/dialogs/unit_list.cpp b/src/gui/dialogs/unit_list.cpp
index 7ac7d77c7cd..31325bdbb58 100644
--- a/src/gui/dialogs/unit_list.cpp
+++ b/src/gui/dialogs/unit_list.cpp
@@ -128,7 +128,7 @@ void unit_list::pre_show(window& window)
 		std::stringstream exp_str;
 		exp_str << font::span_color(unit->xp_color());
 		if(unit->can_advance()) {
-			exp_str << unit->experience() << "/" << unit->max_experience();
+			exp_str << unit->experience_display_for_UI();
 		} else {
 			exp_str << font::unicode_en_dash;
 		}
diff --git a/src/gui/dialogs/unit_recall.cpp b/src/gui/dialogs/unit_recall.cpp
index f2f070d0336..14df1e76f57 100644
--- a/src/gui/dialogs/unit_recall.cpp
+++ b/src/gui/dialogs/unit_recall.cpp
@@ -206,7 +206,7 @@ void unit_recall::pre_show(window& window)
 		std::stringstream exp_str;
 		exp_str << font::span_color(unit->xp_color());
 		if(unit->can_advance()) {
-			exp_str << unit->experience() << "/" << unit->max_experience();
+			exp_str << unit->experience_display_for_UI();
 		} else {
 			exp_str << font::unicode_en_dash;
 		}
diff --git a/src/gui/widgets/unit_preview_pane.cpp b/src/gui/widgets/unit_preview_pane.cpp
index 23f278151e9..ea59c51a298 100644
--- a/src/gui/widgets/unit_preview_pane.cpp
+++ b/src/gui/widgets/unit_preview_pane.cpp
@@ -480,7 +480,7 @@ void unit_preview_pane::set_displayed_unit(const unit& u)
 
 		str << font::span_color(u.xp_color()) << _("XP: ");
 		if(u.can_advance()) {
-			str << u.experience() << "/" << u.max_experience();
+			str << u.experience_display_for_UI();
 		} else {
 			str << font::unicode_en_dash;
 		}
@@ -492,7 +492,7 @@ void unit_preview_pane::set_displayed_unit(const unit& u)
 
 	if(tree_details_) {
 		tree_details_->clear();
-		const std::string unit_xp = u.can_advance() ? (formatter() << u.experience() << "/" << u.max_experience()).str() : font::unicode_en_dash;
+		const std::string unit_xp = u.can_advance() ? (formatter() << u.experience_display_for_UI()).str() : font::unicode_en_dash;
 		tree_details_->add_node("hp_xp_mp", {
 			{ "hp",{
 				{ "label", (formatter() << "<small>" << font::span_color(u.hp_color()) << "<b>" << _("HP: ") << "</b>" << u.hitpoints() << "/" << u.max_hitpoints() << "</span>" << " | </small>").str() },
diff --git a/src/reports.cpp b/src/reports.cpp
index 5c209ad515d..7f0226cf23f 100644
--- a/src/reports.cpp
+++ b/src/reports.cpp
@@ -530,7 +530,7 @@ static config unit_xp(const unit* u)
 	std::ostringstream str, tooltip;
 	str << span_color(u->xp_color());
 	if(u->can_advance()) {
-		str << u->experience() << '/' << u->max_experience();
+		str << u->experience_display_for_UI();
 	} else {
 		str << font::unicode_en_dash;
 	}
diff --git a/src/units/unit.cpp b/src/units/unit.cpp
index fc3070dc8d7..6223856e2d2 100644
--- a/src/units/unit.cpp
+++ b/src/units/unit.cpp
@@ -1008,6 +1008,22 @@ std::string unit::big_profile() const
 	return absolute_image();
 }
 
+std::string unit::experience_display_for_UI() const
+{
+	if (preferences::get("jos_override_experience_display_for_UI", false)) {
+		std::ostringstream out;
+		out << experience();
+		out << "/";
+		out << max_experience();
+		return out.str();
+	} else {
+		std::ostringstream out;
+		out << (max_experience() - experience());
+		out << "↓";
+		return out.str();
+	}
+}
+
 std::string unit::small_profile() const
 {
 	if(!small_profile_.empty() && small_profile_ != "unit_image") {
diff --git a/src/units/unit.hpp b/src/units/unit.hpp
index 2e9fa99fb89..7e45ce5738e 100644
--- a/src/units/unit.hpp
+++ b/src/units/unit.hpp
@@ -441,6 +441,9 @@ public:
 		return max_experience_;
 	}
 
+	/// josteph local patch: Return "42↓", or "0/42XP" if a hidden preference is set.
+	std::string experience_display_for_UI() const;
+
 	/** The number of experience points this unit needs to level up, or 0 if current XP > max XP. */
 	unsigned int experience_to_advance() const
 	{
```

</details>

@ProditorMagnus @vn971 @sevu @soliton-

**Comment by shikadiqueen** [comment]
Personally I would appreciate a more elaborate explanation as to why this is needed.

**Comment by soliton-** [comment]
I would prefer to find a way to show current + remaining XP.

IMO the idea is to remove the rarely needed info about max_experience. I agree that it's rarely needed and it can still be looked up in the help if wanted.

A big question is how to show that in the UI so that it's intuitively understood though...

**Comment by jostephd** [comment]
> IMO the idea is to remove the rarely needed info about max_experience. I agree that it's rarely needed and it can still be looked up in the help if wanted.

The max_experience should also be shown in the tooltip of the blue text. That will show the specific unit's max_experience even if the unit has objects that modify its max_experience.

**Comment by jyrkive** [comment]
My two cents: all games I'm aware of which have player/unit XP show _existing_ XP instead of remaining XP to next level. Hence, showing remaining XP is extremely confusing and a bad idea.

**Comment by CelticMinstrel** [comment]
I'm pretty sure this could be implemented as a `[modification]` for people who prefer it this way.

**Comment by jostephd** [comment]
@CelticMinstrel For the sidebar, I think you're right (https://wiki.wesnoth.org/LuaWML/Display#wesnoth.theme_items), but the patch changes the recruit and recall dialogs too.

For what it's worth, I don't feel strongly about this idea. If it is rejected I'll just keep it as a local patch.

**Comment by jostephd** [comment]
Out of curiosity I looked into @CelticMinstrel's idea, this is what I have so far:

<details>

<summary>
Skeleton of a mod
</summary>

```
local function unit_xp(u)
    local text = tostring(u.max_experience - u.experience) .. "↓"
    return { wml.tag.element {
        image = "attacks/greatsword-elven.png",
        -- See unit::xp_color() for the color value
        text = "xxx<span color='#00A0FF'>" .. text .. "</span>",
        tooltip = "..."
    }}
end

function wesnoth.theme_items.unit_xp()
    local u = wesnoth.get_displayed_unit()
    if not u then return {} end
    return unit_xp(u)
end

function wesnoth.theme_items.selected_unit_xp()
    if wesnoth.get_selected_tile() then
        local u = wesnoth.get_unit(wesnoth.get_selected_tile())
        if not u then return {} end
        return unit_xp(u)
    end
end
```

</details>

Issues:

- [ ] Works in 1.14 but not in master (regression in master?)
- [ ] Doesn't affect recruit dialog, recall dialog, unit preview pane, etc
- [ ] Need to reimplement the coloring (blue at start, light blue/white near advancement, purple if unit is max level)
- [ ] Loses the "Experience Factor: 100%" information from the tooltip

**Comment by sevu** [comment]
The advantage I see in this is that there is less to keep in mind when planing attacks.

There are many things to calculate when planning an attack, which unit hits better, how's the chance that 2 units are enough for the attack... if the chance is high, can I use a weaker unit instead and try to level it... what if things go wrong?

_How many XP does the unit xy has left?_ Does it need need 1 kill + 3 fights, or 1 kill + 5 fights?
The XP are quite low in the list of things which I take into account, but are also quite important too.
It takes time to look at things. Nobody likes waiting too long for the opponent in MP.

There's already a full scale combat simulator, some help on the XP part by having to calculate one step less would be an improvement.

**Comment by gfgtdf** [comment]
There are other cases than ravannas that use max_xp, for example in my mod i have a pickupable item with a simple `apply_to=max_experience increase=-50%` effect that permementyl decreases its max=xpand so giving it to a unit with 10/50 would onyl resuce its needed xp to 15, but giving it to a unit with 60/100 would make it advance with 10xp left.

Also iirc there is a unit in a mainline scenario that gives its xp to other adjacent units.

**Comment by vgaming** [comment]
> all games I'm aware of which have player/unit XP show existing XP instead of remaining XP to next level.

The first reaction is that this statement is true. Thinking further about it though -- is it? Many games have e.g. thousands/millions of, say, "experience", and it is important to show the user that their slow progress does step-by-step lead to their, say, level advance.

There are other scenarios though. What if you have e.g. a caravan of items that goes from city A to city B? Many users prefer to show you "3 days left" instead of "4 days passed out of 7". Same with some online games where you can command your hero to move from A to B. It usually says "will reach town B in 17 minutes", not "17 minutes out of 34 have passed".

To sum up, I think we're good to go from a "least surprise" principle. Presuming "remaining XP" is what users really want to see.

**Comment by vgaming** [comment]
@gfgtdf well to be fair, you _could_ decrease both current XP left to advance and maximum XP? E.g. you had 10 XP to advance and 50 maximum -- after picking up the item you'll have 5 XP left to advance and 25 maximum.
Harder to implement, but reasonable for UI.

**Comment by vgaming** [comment]
EDIT ^

**Comment by CelticMinstrel** [comment]
> > all games I'm aware of which have player/unit XP show existing XP instead of remaining XP to next level.
> 
> The first reaction is that this statement is true. Thinking further about it though -- is it?
I can definitely think of games that show the XP needed to reach the next level. I think it's usually _in addition to_ current total XP - so, you see the amount of XP the character currently possesses, as well as the additional amount required to reach the next level. I can't think of any that instead show the _total_ amount at which you'll reach the next level (though that doesn't mean they don't exist).

**Comment by jostephd** [comment]
> @gfgtdf well to be fair, you _could_ decrease both current XP left to advance and maximum XP? E.g. consider that you had 10 XP to advance and 50 maximum. After picking up the item you'll have 5 XP left to advance and 25 maximum.
> Harder to implement, but reasonable for UI.

I think you've put the cart before the horse. The UI should be designed to fit the needs of gameplay, not the other way around.

<details><summary>More about @gfgtdf's item</summary>

Furthermore, @vgaming, the semantics you propose are very different to @gfgtdf's. In his version, the unit starts out with 40/50XP, then it changes to 40/25XP so it advances and has 40-25=15XP left over. In your version, the unit starts with 40/50XP and is then halved to 20/25XP. The net effect in gfgtdf's version is a 25XP _bonus_ and in your version a 20XP _malus_.

What would be closer is if gfgtdf described his item as "Gives the unit an XP bonus equal to half that unit's maximum XP". For example, a Red Mage with 10XP left to advance would receive a bonus of 50XP. Under the default rules this is equivalent (but it's not equivalent if somebody picks up gfgtdf's item in a game that has Ravana's XP mod enabled...)

</details>

By the way, does anyone have another proposal for what the UI might be? That is, what else might the sidebar read besides "XP: 40/50" (current behavior) and "XP: 10↓" (my proposal).

**Comment by jostephd** [comment]
> Out of curiosity I looked into @CelticMinstrel's idea, this is what I have so far:
> Skeleton of a mod

I wasn't planning to work on the lua version any further, by the way. If somebody wants to take the code and add the missing coloring logic and package it as an addon, be my guest...

**Comment by jostephd** [comment]
The "accumulated XP" is also important when trying to assess whether a unit would be worthy of recall. For example, near the end of a scenario a player might be more willing to sacrifice a unit with 4XP accumulated than one with 20XP accumulated.

**Comment by ProditorMagnus** [comment]
Now that comment goes against the base premise of this issue.

**Comment by jostephd** [comment]
I don't think so. That comment is not an argument against showing "remaining XP" in the UI; it is an argument for showing "accumulated XP" in the UI. The UI could show both. Or it could show one of the figures in the sidebar and the other in a tooltip. Or it could show "40/50XP" and change to "10↓" when the player presses the blue XP text with the mouse (this idea is CelMin's, not mine). And so on, that was just a tiny bit of brainstorming...

**Comment by vgaming** [comment]
@jostephd I'm well aware that what I'm proposing is behavior change. I did it explicitly. There are different ways how you can reduce XP requirements by a factor of 2, with a one-time bonus or without it. What I'm saying is that the lack of a one-time bonus looks like a sane schema for me as a user as well.

> The UI should be designed to fit the needs of gameplay, not the other way around.

Of course. Gameplay-wise, the behaviour I proposed sounds even a bit MORE intuitive for me. Like if you had N% to advance, you retain it, but you move 2 times faster towards your goal from now on. Anyway, it's subjective and depends on the map/game heavily. Only thing I'm saying is it can be an explicit alternative.

**Comment by jostephd** [comment]
Pre-1.15.0 triage: What do we do with this issue? I guess we could close it, postpone it, or add a preference for choosing between the "10/100" style and the "10↓ 90↑" style.

**Comment by jostephd** [comment]
I wonder if we can do it without a preference. Maybe show all three numbers? For example, we could change the sidebar label from just "XP" to "XP needed" (or something) and then say `90 (10/100)`.

**Comment by ProditorMagnus** [comment]
Tooltip should show all 3 values, I remember from SX that even with current 2, if values go to 3+4 digits then part is not visible.

**Comment by jostephd** [comment]
> even with current 2, if values go to 3+4 digits then part is not visible.

That's just a bug in the default theme. After changing `[unit_xp]rect` to `"=,+0,=+30,+14"` I get:

![2019-09-14-072339_173x146_scrot](https://user-images.githubusercontent.com/24784687/64904938-bcbbcd00-d6c0-11e9-8b2f-7b4dc64052fb.png)

(the label is truncated, that's easy to fix though)

**Comment by vgaming** [comment]
I think two numbers are less confusing. E.g. 100 / 500, as we have now, just the meaning of the first number and the tooltip are swapped.

**Comment by ProditorMagnus** [comment]
Does wesnoth support changing display format when you click the xp value?

**Comment by Wedge009** [comment]
Oh, you mean toggling the display format just by clicking (instead of having to go into preferences)? That would be really nice if possible.

**Comment by jostephd** [comment]
> Oh, you mean toggling the display format just by clicking (instead of having to go into preferences)? That would be really nice if possible.

Clicks on the blue text are handled here:
https://github.com/wesnoth/wesnoth/blob/de8a9133745a88840d02fb0a014f0afdd6624da3/src/mouse_events.cpp#L805

(I just happened to find this while looking at #1297) 

Anyway, I imagine it's implementable. (And then the recruit dialog, unit list dialog, etc could follow the style the sidebar uses, I guess)

**Comment by jostephd** [comment]
I'm not sure about clicking on the label to change its appearance: it's elegant once you know about it, sure, but I don't think it'd be easy to discover... :thinking:

**Comment by Wedge009** [comment]
Label and/or value would be good. It might be a 'hidden' feature (like the report about zoom in #1213) but I think it's better having it than not. Also we can update the in-game Help for the side-bar interface to document the feature.

**Comment by vgaming** [comment]
We can also have the toggle in both places, right? On the main screen accepting mouse clicks, and in Preferences in Advanced section.

BTW, if we'll start accepting mouse clicks, we can change the hover to include `click to toggle display` as well. So you can more easily discover the toggle while you hover various things on the panel.

**Comment by jostephd** [comment]
@vgaming Yes, we can have the toggle in both places, and we can say in the tooltip that clicking the label changes its display. (We'd have to implement that "click to change" thing first, though. I'm not sure how much work that would be. Also, we'd have to choose what formats to cycle between, possibly more than two of them.)

**Comment by cooljeanius** [comment]
existing XP display is perfectly fine for me; this seems like it would just make things more complicated and harder to understand

**Comment by Wedge009** [comment]
I think as said above I also think the _display_ in the side-bar should be left as-is. However, there are still places where the XP-to-advance figure would be useful. The main places for me would be the unit list and the unit recall dialogues, mainly for knowing what units are closest to advancing but also for considering what units might be targeted by a computer opponent (since the default AI tends to focus on attacking those as priority). This came out of the discussion of #1294 as well, but I'm not sure if this issue is focused solely on the side-bar implementation or could consider those aforementioned dialogues instead or in addition to.
